bp-adv-diff

Raamkrishnan opened the pull request:
<!--
This template should be included in all pull requests. Items in the list should
either be completed by the original author or explicitly dismissed by one of the
IBAMR principal developers.

IBAMR is a community effort and it wouldn't exist without people contributing
code. Thanks in advance for helping to make IBAMR better!
-->

### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [ ] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [ ] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [ ] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [ ] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [ ] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

I have written a new subclass named ```BrinkmanAdvDiffSemiImplicitHierarchyIntegrator``` and do all the necessary modifications required to implement a penalized advection-diffusion equation. The additional diffusion coefficient and the forcing terms required to impose flux boundary conditions on the embedded surface is obtained from ```BrinkmanPenalizationAdvDiff``` class. I have also added two more examples in ```examples/adv_diff``` folder.

Raamkrishnan commented:
I have tried to rebase against master and rewrite history. But I was getting conflict in ```lib/Makefile.in``` which I am not sure how to resolve.

amneetb commented:
@Raamkrishnan : Can you rename `BrinkmanPenalizationAdvDiff` to `BrinkmanAdvDiffBcHelper` to be consistent with the integrator class?

amneetb commented:
`registerBrinkmanPenalization()` routine name to `registerBrinkmanAdvDiffBcHelper()`?

boyceg commented:
I have some time sensitive stuff to get done this week, but will aim to take a look by early next week at the latest.

boyceg commented:
> I have tried to rebase against master and rewrite history. But I was getting conflict in `lib/Makefile.in` which I am not sure how to resolve.

These generated files can be problematic when rebasing. What you can do is to re-run `bootstrap` after resolving any other conflicts, and then commit the updated generated files.

Note that we have _tried_ to set up `bootstrap` and `configure` to ensure that everyone automatically uses the same versions of `autoconf`, `automake`, and `libtool` --- so doing this should not create a huge number of changes to all of the generated files.

boyceg commented:
> > I have tried to rebase against master and rewrite history. But I was getting conflict in `lib/Makefile.in` which I am not sure how to resolve.
> 
> These generated files can be problematic when rebasing. What you can do is to re-run `bootstrap` after resolving any other conflicts, and then commit the updated generated files.
> 
> Note that we have _tried_ to set up `bootstrap` and `configure` to ensure that everyone automatically uses the same versions of `autoconf`, `automake`, and `libtool` --- so doing this should not create a huge number of changes to all of the generated files.

@drwells, I know that you are hoping to be done with autotools soon 😁, but is there a good place to document possible workflows for dealing with these kinds of conflicts when rebasing?

Raamkrishnan commented:
Sorry, the last commit that I pushed has some compilation errors. I thought I tested the examples too. But I think I did something wrong while incorporating Amneet's suggestions. Let me correct that and I will open PR again.

drwells commented:
It's totally fine to keep a PR open that isn't ready yet - just let us know and we can mark it as WIP.

drwells commented:
> @drwells, I know that you are hoping to be done with autotools soon :grin:, but is there a good place to document possible workflows for dealing with these kinds of conflicts when rebasing?

I think we can just keep suffering with manual suggestions for a little longer - these conflicts usually require manual intervention anyway.

Raamkrishnan commented:
@drwells I have pushed the code and also tested the examples and the test. You can review the code now.

amneetb commented:
@drwells : Ping :-)

boyceg commented:
I think this one is mostly my job to review. Unfortunately I am currently buried in some other stuff.

drwells commented:
Sorry about the delay - I was out of town for the last half week. I meant to tackle this today but didn't have time to do it. I will dig deep tomorrow.

drwells commented:
Fixes #627.

drwells reviewed (changes requested):
Looks good - I have a few minor requests but this resolves my main concern from the last round of reviews so I'm confident that we can get this merged soon.

The Brinkman implementation looks reasonable to me but perhaps someone who is more familiar with this stuff should also sign off on the implementation.

drwells commented:
One more thing - we should definitely add a major changelog entry so that you get credit for all this work. Do you know how to do that?

Raamkrishnan commented:
Thank @drwells. I will make all the requested changes tomorrow. Yes. I know how to add a changelog entry.

Raamkrishnan commented:
@drwells I have incorporated your suggestions. In addition to the earlier code,I have added modifications to handle Robin boundary conditions in BrinkmanAdvDiffBcHelper class.  Therefore, the present Brinkman code can handle Dirichlet, Neumann and Robin bcs for advection-diffusion equation. Some more examples also I added in `examples/adv_diff` directory. Please take a look.

drwells commented:
Awesome. I'll dig in and review this tomorrow.

drwells commented on tests/adv_diff/bp_adv_diff.cpp:
```diff
@@ -0,0 +1,413 @@
+
+void
+evaluate_brinkman_bc_callback_fcn(int B_idx,
+                                  Pointer<CellVariable<NDIM, double> > ls_var,
```
```suggestion
                                  Pointer<CellVariable<NDIM, double> > /*ls_var*/,
```

drwells commented on tests/adv_diff/bp_adv_diff.cpp:
```diff
@@ -0,0 +1,413 @@
+                                  Pointer<CellVariable<NDIM, double> > ls_var,
+                                  Pointer<HierarchyMathOps> hier_math_ops,
+                                  double /*time*/,
+                                  void* ctx)
```
```suggestion
                                  void* /*ctx*/)
```

drwells reviewed (changes requested):
I suggested a few minor changes, including one fix.

drwells commented:
I am pretty sure this class doesn't work with restarts since time dependence isn't set up correctly - we also don't have any parallel tests. Do these things work? If so, are you willing to add tests for them?

We also still need a changelog entry.

amneetb commented on src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp:
```diff
@@ -0,0 +1,883 @@
+#endif
+    if (!is_from_restart)
+    {
+        AdvDiffSemiImplicitHierarchyIntegrator::getFromInput(db, is_from_restart);
```
@Raamkrishnan : just remove the BrinkmanAdvDiffSemiImplicitHierarchyIntegrator::getFromInput() routine and in the constructor itself read in the brinkman_time_independent option (without checking if it restarted run or not)

When a simulation restarts, it still reads in the input file. Only those quantities which maintain a state in the restart database are not read from the input file, but for most of the input file options, we do not maintain their state in the restart database. We prefer to read them again from the input file.

amneetb commented on src/IB/BrinkmanAdvDiffBcHelper.cpp:
```diff
@@ -0,0 +1,974 @@
+BrinkmanAdvDiffBcHelper::registerHomogeneousBC(Pointer<CellVariable<NDIM, double> > Q_var,
+                                               Pointer<CellVariable<NDIM, double> > ls_solid_var,
+                                               std::string bc_type,
+                                               std::string indicator_func_type,
```
@drwells : which value?

drwells replied:
`indicator_func_type`.

amneetb commented on src/IB/BrinkmanAdvDiffBcHelper.cpp:
```diff
@@ -0,0 +1,974 @@
+                    // Homogeneous Neumann BCs have no contribution to the forcing term
+                    if (bc_type == DIRICHLET)
+                    {
+                        brinkman_forcing += (chi / eta * bc_val);
```
@Raamkrishnan : for better readability use parentheses at the correct spot 
`brinkman_forcing += (chi / eta) * bc_val;`

amneetb commented on src/IB/BrinkmanAdvDiffBcHelper.cpp:
```diff
@@ -0,0 +1,974 @@
+                        // Note: assumes that chi is positive when phi is negative
+                        const double chi = (1.0 - Hphi);
+
+                        (*F_data)(ci) += chi / eta * (*g_data)(ci);
```
Same here 

` (*F_data)(ci) += (chi / eta) * (*g_data)(ci);`

amneetb reviewed (commented):
> I am pretty sure this class doesn't work with restarts since time dependence isn't set up correctly - we also don't have any parallel tests. Do these things work? If so, are you willing to add tests for them?

Yes, things work in parallel --- we have tested codes on clusters. Rama can convert one of the serial tests to parallel tests. 
 
> 
> We also still need a changelog entry.

drwells commented:
As long as we are using uniform grids in serial we should get exactly the same solution (and therefore whatever computed values) in parallel so hopefully this is easy to add and check.

drwells commented on src/IB/BrinkmanAdvDiffBcHelper.cpp:
```diff
@@ -0,0 +1,974 @@
+    void* ctx = nullptr;
+
+    // Get the user specified indicator function.
+    auto indicator_func_type_enum = string_to_enum<IndicatorFunctionType>(indicator_func_type);
```
Perhaps I misread the compiler warning - I'll check.

Raamkrishnan commented:
@drwells @amneetb Thanks for your review. I have incorporated those changes and pushed it. I have added a parallel test as well. Please take a look and let me know your suggestions.

Raamkrishnan commented:
@drwells It would be better if we merge this branch and #1139 branch to the master. Because the code that I am developing now requires both of these changes. Please take a look on these branches and let me know your comments. For #1139 branch, I have to add some more tests and  I am working on that now.

drwells commented:
Can do. I am traveling today but I will find time to look at this and the other PR either tonight or tomorrow.

drwells reviewed (approved):
Sorry for the delay. The changes I requested were pretty small and they've all been made so this looks ready (once we fix the merge conflict).

Raamkrishnan commented:
@drwells I have resolved the merge conflict. Before pushing, I have also tested the examples and the tests that I have added in this branch and everything is working fine. Please take a look.

drwells commented:
Looks good. I will apply the CMake changes in #1184 directly to this PR and then merge.

amneetb commented:
Thanks @drwells !
